### PR TITLE
Best practices for regular expressions versus `Regex` performance review

### PR DESCRIPTION
**Pull request by EngRajabi**
## Proposed changes

`Regex` improvements:
  - set option compile
  - set timeout
  - set static object (`Regex` class is immutable)

## Links
- Microsoft Docs: [Best Practices for Regular Expressions in .NET | Microsoft Learn](https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices)
- .NET Blog: [Regular Expression Improvements in .NET 7](https://devblogs.microsoft.com/dotnet/regular-expression-improvements-in-dotnet-7/)

### CODE REVIEW

**Review by Hossein-Edraki — APPROVED** [review]
It's Ok

**Comment by EngRajabi** [comment]
@TomPallister

**Review by VahidTajari — DISMISSED** [review]

**Comment by WeihanLi** [comment]
What if timeout happens? Timeout maybe not good practice here.

**Comment by EngRajabi** [comment]
Setting timeout is one of the necessary tasks of using regex. When a timeout occurs, it is best to run the entire project involved and not respond to others.

**Comment by WeihanLi** [comment]
Is there some reason for considering `100ms` as the regex timeout, and how could the client update the timeout config if the client wanna use a longer timeout?

**Comment by EngRajabi** [comment]
timeout is not customizable by the customer. Like many other settings. It was 100 ms when I adjusted it according to the tests I had. But we can decide here and change this time. We can set this time to 500 ms. If we do not receive an answer at this time, it is better not to involve the whole project in its implementation and let it respond to other requests.

**Review by EngRajabi — COMMENTED** [review]
review done

**Comment by raman-m** [comment]
Dear Mohsen,

Why did you [merge develop branch from your fork into the feature branch](#commits-pushed-ff1dbb6)?
I meant these commits: https://github.com/ThreeMammals/Ocelot/pull/1348/commits/ff1dbb6c1d9925a08e910c7c4d1f79fd6017c216 https://github.com/ThreeMammals/Ocelot/pull/1348/commits/5b336cdda05a6474eb964b5668272ed6f3b358df
It was unnecessary because [I did the merge of ThreeMammals:develop](#commits-pushed-329d2c3) in https://github.com/ThreeMammals/Ocelot/pull/1348/commits/329d2c373dfbd7619d2d570a77327bb6d3b02268.
After that I trigger pipeline to make a valid green build for commit https://github.com/ThreeMammals/Ocelot/pull/1348/commits/03a70b32b23f00f41ab93d14e7bbf75a775066a3.

Now there is no build at all and pipeline cannot run new build.
The pipeline cannot recognize **develop** branch because you have juggled all commits in a wrong way.

Could you [fix/update your fork develop branch](/EngRajabi/Ocelot/pull/8#issuecomment-1643837790) please first?

After that we will try to fix pipeline triggering by rebasing current feature branch onto **develop** one.

**Review by EngRajabi — COMMENTED** [review]
I did Branch Develop on Branch as Rebase. Does this solve this problem?

**Comment by raman-m** [comment]
image

Now I see develop and it has [zero diff](../compare/develop...EngRajabi:Ocelot:develop), but the branch is not default. ☝️ 
You have to go to [repo settings](/EngRajabi/Ocelot/settings) and make **develop** branch as default!
I hope that should help to fix the problem of builds, because all builds are built for develop branch (auto) and master (manually).
Currently I don't see that pipeline was triggered for [top commit](/EngRajabi/Ocelot/commits/improvement_regex).
Maybe an empty or tiny commit will be required to trigger the pipeline after making develop branch default.

**Review by EngRajabi — COMMENTED** [review]
change something for build

**Comment by raman-m** [comment]
Feature branch has been rebased onto develop!
The build has been passed!

Going to start code review...

**Review by RaynaldM — APPROVED** [review]

**Review by wast — APPROVED** [review]

File: src/Ocelot/Configuration/Validator/FileConfigurationFluentValidator.cs [hunk 1]
```diff
@@ -107,8 +104,7 @@ private static bool AllRoutesForAggregateExist(FileAggregateRoute fileAggregateR
 
         private static bool IsPlaceholderNotDuplicatedIn(string upstreamPathTemplate)
         {
-            var regExPlaceholder = new Regex("{[^}]+}");
-            var matches = regExPlaceholder.Matches(upstreamPathTemplate);
+            var matches = _regExPlaceholder.Matches(upstreamPathTemplate);
```
Comment (raman-m) [inline src/Ocelot/Configuration/Validator/FileConfigurationFluentValidator.cs:107]: No catching and processing of the `RegexMatchTimeoutException` exception.

Will it result to unhandled exception by app domain?
Wil it be caught by upper context?

File: src/Ocelot/Authorization/ClaimsAuthorizer.cs [hunk 2]
```diff
@@ -38,7 +37,7 @@ List<PlaceholderNameAndValue> urlPathPlaceholderNameAndValues
                 if (values.Data != null)
                 {
                     // dynamic claim
-                    var match = Regex.Match(required.Value, @"^{(?<variable>.+)}$");
+                    var match = _regexAuthorize.Match(required.Value);
```
Comment (raman-m) [inline src/Ocelot/Authorization/ClaimsAuthorizer.cs:40]: No catching and processing of the `RegexMatchTimeoutException` exception.

Will it result to unhandled exception by app domain?
Wil it be caught by upper context?

File: src/Ocelot/Configuration/Validator/RouteFluentValidator.cs [hunk 3]
```diff
@@ -1,19 +1,21 @@
...
 
 namespace Ocelot.Configuration.Validator
 {
     public class RouteFluentValidator : AbstractValidator<FileRoute>
     {
         private readonly IAuthenticationSchemeProvider _authenticationSchemeProvider;
+        private static readonly Regex _secondsRegEx = new Regex("^[0-9]+s", RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
+        private static readonly Regex _minutesRegEx = new Regex("^[0-9]+m", RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
+        private static readonly Regex _hoursRegEx = new Regex("^[0-9]+h", RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
+        private static readonly Regex _daysRegEx = new Regex("^[0-9]+d", RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
```
Comment (raman-m) [inline src/Ocelot/Configuration/Validator/RouteFluentValidator.cs:18]: Hardcode has detected!
The `100` milliseconds constant should be public static.

File: src/Ocelot/Configuration/Validator/RouteFluentValidator.cs [hunk 4]
```diff
@@ -116,15 +118,10 @@ private static bool IsValidPeriod(FileRateLimitRule rateLimitOptions)
...
-            var daysRegEx = new Regex("^[0-9]+d");
-
-            return secondsRegEx.Match(period).Success
-                   || minutesRegEx.Match(period).Success
-                   || hoursRegEx.Match(period).Success
-                   || daysRegEx.Match(period).Success;
+            return _secondsRegEx.Match(period).Success
+                   || _minutesRegEx.Match(period).Success
+                   || _hoursRegEx.Match(period).Success
+                   || _daysRegEx.Match(period).Success;
```
Comment (raman-m) [inline src/Ocelot/Configuration/Validator/RouteFluentValidator.cs:124]: No catching and processing of the `RegexMatchTimeoutException` exception.

Will it result to unhandled exception by app domain?
Wil it be caught by upper context?

File: src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs [hunk 5]
```diff
@@ -1,24 +1,25 @@
...
 namespace Ocelot.DependencyInjection
 {
     public static class ConfigurationBuilderExtensions
     {
+        private static readonly Regex _reg = new Regex(@"^ocelot\.(.*?)\.json$", RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
+
         [Obsolete("Please set BaseUrl in ocelot.json GlobalConfiguration.BaseUrl")]
         public static IConfigurationBuilder AddOcelotBaseUrl(this IConfigurationBuilder builder, string baseUrl)
         {
+            Regex.CacheSize += 100;
```
Comment (raman-m) [inline src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs:21]: What is the default, current value of the cache in runtime?
So, why did you increase this property value?

File: src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs [hunk 6]
```diff
@@ -39,19 +40,17 @@ public static IConfigurationBuilder AddOcelot(this IConfigurationBuilder builder
 
         public static IConfigurationBuilder AddOcelot(this IConfigurationBuilder builder, string folder, IWebHostEnvironment env)
         {
+            Regex.CacheSize += 100;
```
Comment (raman-m) [inline src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs:43]: Do you consider the `AddOcelot` methos as an entry point for entire Ocelot pipeline, and for all regex-objects of the app domain?

File: src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs [hunk 7]
```diff
@@ -39,19 +40,17 @@ public static IConfigurationBuilder AddOcelot(this IConfigurationBuilder builder
...
-            const string subConfigPattern = @"^ocelot\.(.*?)\.json$";
-
             var excludeConfigName = env?.EnvironmentName != null ? $"ocelot.{env.EnvironmentName}.json" : string.Empty;
 
-            var reg = new Regex(subConfigPattern, RegexOptions.IgnoreCase | RegexOptions.Singleline);
-
             var files = new DirectoryInfo(folder)
                 .EnumerateFiles()
-                .Where(fi => reg.IsMatch(fi.Name) && (fi.Name != excludeConfigName))
+                .Where(fi => _reg.IsMatch(fi.Name) && (fi.Name != excludeConfigName))
```
Comment (raman-m) [inline src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs:53]: No catching and processing of the `RegexMatchTimeoutException` exception.

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs [hunk 8]
```diff
@@ -102,20 +98,25 @@ public async Task Invoke(HttpContext httpContext)
             await _next.Invoke(httpContext);
         }
 
-        private static void RemoveQueryStringParametersThatHaveBeenUsedInTemplate(DownstreamRequest downstreamRequest, List<PlaceholderNameAndValue> templatePlaceholderNameAndValues)
+        private static void RemoveQueryStringParametersThatHaveBeenUsedInTemplate(
+            DownstreamRequest downstreamRequest, List<PlaceholderNameAndValue> templatePlaceholderNameAndValues)
```
Comment (raman-m) [inline src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:102]: Why was that changed? 😉 
There is no need to change! No new code!

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs [hunk 9]
```diff
@@ -102,20 +98,25 @@ public async Task Invoke(HttpContext httpContext)
...
                 {
                     var questionMarkOrAmpersand = downstreamRequest.Query.IndexOf(name, StringComparison.Ordinal);
                     downstreamRequest.Query = downstreamRequest.Query.Remove(questionMarkOrAmpersand - 1, 1);
 
-                    var rgx = new Regex($@"\b{name}={nAndV.Value}\b");
-                    downstreamRequest.Query = rgx.Replace(downstreamRequest.Query, string.Empty);
+                    var pattern = $@"\b{name}={nAndV.Value}\b";
+                    var regex = _regex.AddOrUpdate(pattern,
+                        new Regex(pattern, RegexOptions.Compiled, TimeSpan.FromMilliseconds(100)),
+                        (key, oldValue) => oldValue);
```
Comment (raman-m) [inline src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:117]: No catching and processing of the `RegexMatchTimeoutException` exception.

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs [hunk 10]
```diff
@@ -125,22 +126,24 @@ private static void RemoveQueryStringParametersThatHaveBeenUsedInTemplate(Downst
...
-            return dsPath.Value.Substring(dsPath.Value.IndexOf('?', StringComparison.Ordinal));
+            return dsPath.Value.Substring(dsPath.Value.IndexOf('?', StringComparison.Ordinal));
         }
 
-        private static bool ContainsQueryString(DownstreamPath dsPath)
+        private static bool ContainsQueryString(DownstreamPath dsPath)
         {
-            return dsPath.Value.Contains('?');
+            return dsPath.Value.Contains('?');
         }
```
Comment (raman-m) [inline src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:142]: Still wonder...
What are the changes? 🤣

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs [hunk 11]
```diff
@@ -125,22 +126,24 @@ private static void RemoveQueryStringParametersThatHaveBeenUsedInTemplate(Downst
...
 
-        private (string Path, string Query) CreateServiceFabricUri(DownstreamRequest downstreamRequest, DownstreamRoute downstreamRoute, List<PlaceholderNameAndValue> templatePlaceholderNameAndValues, Response<DownstreamPath> dsPath)
+        private (string Path, string Query) CreateServiceFabricUri(DownstreamRequest downstreamRequest,
+            DownstreamRoute downstreamRoute, List<PlaceholderNameAndValue> templatePlaceholderNameAndValues,
+            Response<DownstreamPath> dsPath)
```
Comment (raman-m) [inline src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:146]: I see zero changes in method signature!
But, If you split parameters, I suggest to put each parameter onto new line 👇 
```suggestion
        private (string Path, string Query) CreateServiceFabricUri(
            DownstreamRequest downstreamRequest,
            DownstreamRoute downstreamRoute,
            List<PlaceholderNameAndValue> templatePlaceholderNameAndValues,
            Response<DownstreamPath> dsPath)
```

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs [hunk 12]
```diff
@@ -150,7 +153,8 @@ private static bool ContainsQueryString(DownstreamPath dsPath)
 
         private static bool ServiceFabricRequest(IInternalConfiguration config, DownstreamRoute downstreamRoute)
         {
-            return config.ServiceProviderConfiguration.Type?.ToLower() == "servicefabric" && downstreamRoute.UseServiceDiscovery;
+            return config.ServiceProviderConfiguration.Type?.ToLower() == "servicefabric" &&
+                   downstreamRoute.UseServiceDiscovery;
```
Comment (raman-m) [inline src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:157]: Zero changes!
Revert this fake change plz!

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs [hunk 13]
```diff
@@ -91,7 +86,8 @@ public async Task Invoke(HttpContext httpContext)
                 }
                 else
                 {
-                    RemoveQueryStringParametersThatHaveBeenUsedInTemplate(downstreamRequest, templatePlaceholderNameAndValues);
+                    RemoveQueryStringParametersThatHaveBeenUsedInTemplate(downstreamRequest,
+                        templatePlaceholderNameAndValues);
```
Comment (raman-m) [inline src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:90]: Revert this!

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs [hunk 14]
```diff
@@ -66,11 +60,12 @@ public async Task Invoke(HttpContext httpContext)
...
             {
-                var pathAndQuery = CreateServiceFabricUri(downstreamRequest, downstreamRoute, templatePlaceholderNameAndValues, response);
+                var pathAndQuery = CreateServiceFabricUri(downstreamRequest, downstreamRoute,
+                    templatePlaceholderNameAndValues, response);
 
                 //todo check this works again hope there is a test..
-                downstreamRequest.AbsolutePath = pathAndQuery.Path;
-                downstreamRequest.Query = pathAndQuery.Query;
+                downstreamRequest.AbsolutePath = pathAndQuery.Path;
+                downstreamRequest.Query = pathAndQuery.Query;
```
Comment (raman-m) [inline src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:68]: Zero changes!
Revert to previous code!

File: src/Ocelot/Infrastructure/ConfigAwarePlaceholders.cs [hunk 15]
```diff
@@ -51,7 +49,7 @@ public Response Remove(string key)
             => _placeholders.Remove(key);
 
         private static string CleanKey(string key)
-            => Regex.Replace(key, @"[{}]", string.Empty, RegexOptions.None);
+            => _regex.Replace(key, string.Empty);
```
Comment (raman-m) [inline src/Ocelot/Infrastructure/ConfigAwarePlaceholders.cs:52]: No catching and processing of the `RegexMatchTimeoutException` exception.

File: src/Ocelot/Values/UpstreamPathTemplate.cs [hunk 16]
```diff
@@ -1,18 +1,24 @@
...
         {
             Template = template;
             Priority = priority;
             ContainsQueryString = containsQueryString;
             OriginalValue = originalValue;
             Pattern = template == null ?
-                new Regex("$^", RegexOptions.Compiled | RegexOptions.Singleline) :
-                new Regex(template, RegexOptions.Compiled | RegexOptions.Singleline);
+                _reg :
+                    _regex.AddOrUpdate(template, new Regex(template, RegexOptions.Compiled | RegexOptions.Singleline, TimeSpan.FromMilliseconds(100)),
```
Comment (raman-m) [inline src/Ocelot/Values/UpstreamPathTemplate.cs:20]: No catching and processing of the `RegexMatchTimeoutException` exception.

File: test/Ocelot.UnitTests/Cache/CacheKeyGeneratorTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Cache/CacheKeyGeneratorTests.cs:1]: There are no unit tests for new approach of using Regex.
Time out should be tested!

File: test/Ocelot.UnitTests/CacheManager/OutputCacheMiddlewareRealCacheTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/CacheManager/OutputCacheMiddlewareRealCacheTests.cs:1]: Regex Time out event should be tested!

**Review by raman-m — CHANGES_REQUESTED** [review]
I see three major issues with current code:
- `TimeSpan.FromMilliseconds(100)`. The hardcode of `100` is detected. Milliseconds constant should be a constant, and public static. Moving to some common options static class is highly desired.
- No catching and processing of the `RegexMatchTimeoutException` object being generated by all `Regex` objects. A `try-catch` constructions should be provided.
- No unit tests which cover new behavior of `Regex`. So, `RegexMatchTimeoutException` exception case should be tested.

Hope this doc will be helpful: [Defining a Time-Out Value](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex?view=net-7.0#define_timeout),
to understand current issues.

**Comment by EngRajabi** [comment]
@raman-m 
Hello, I read some of your reviews, but since there were many, I will answer you in one post.
1- In the case of TimeSpan.FromMilliseconds(100), if the work is not done in 100 milliseconds, the work will not continue. Regarding the RegexMatchTimeoutException, I ask you what happens in the current situation where we have not set the timeout??? When we don't have a timeout, the requests wait so much that they receive errors, and the CPU and computing resources are busy. But with my changes, a timeout error occurs, which probably returns the error code 500 in the output. This behavior is better than the current default behavior.
2- In the case of Regex.CacheSize += 100. Its default value is 15. According to the following document
https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices
This amount is very, very low for large projects and causes the cache to be cleared very often and this causes our patterns to come out of the cache.
3- Regarding format code changes. These changes are not my work. If you have looked, people changed my code.

**Comment by raman-m** [comment]
> @EngRajabi commented [on Aug 4, 2023](#issuecomment-1665463285)

It is better to reply in own thread of each code review issue...

---

>  When we don't have a timeout, the requests wait so much that they receive errors, and the CPU and computing resources are busy. But with my changes, a timeout error occurs, **which probably returns the error code 500 in the output.** This behavior is better than the current default behavior.

Thanks for improvements! Current behavior has been improved. We've done with but...
I'm pretty sure that each `Match` method is not wrapped by any `try-catch` construction in upper contexts. That means, Yes, we have 99% `UnhandledException` which will be handled globally by current [OcelotPipelineExtensions](../blob/develop/src/Ocelot/Middleware/OcelotPipelineExtensions.cs#L40).
So, in this case we have error being handled, request has been failed, and records will be added by Ocelot loggers.
But... Instead of having error logged why not to improve user experience, and let the user know that current timeout value is low, and it is better to increase time out value. For example we could log a warning with all required metrics.
Finally, we can introduce some `Regex` helper for `Match` methods to process timeouts.

---

And, Could you add unit/acceptance test(s) please?

**Comment by raman-m** [comment]
@WeihanLi commented [on Nov 20, 2020](#issuecomment-730959862):
> Is there some reason for considering 100ms as the regex timeout, and how could the client update the timeout config if the client wanna use a longer timeout?

I **wanna** develop your idea further! 😉
The `100ms` value will be OK if it covers all cases of `Regex`-matching inside of Ocelot core. But I am sure it doesn't. And, "[**Houston, we have a problem!**](https://www.youtube.com/watch?v=YwG4F-16Tno&t=40s)"  🤣 The problem with the `RegexMatchTimeoutException` objects being thrown by `Match` methods. 😩 

Regarding configuring of timeout value. Well... We could introduce one common configurable value for this option in [Best Practices for Regular Expressions in .NET](https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices#use-time-out-values):
> - Set a process-wide or AppDomain-wide value with code such as `AppDomain.CurrentDomain.SetData("REGEX_DEFAULT_MATCH_TIMEOUT", TimeSpan.FromMilliseconds(100));`.

What about such kind of `Regex` configuration?

**Review by raman-m — CHANGES_REQUESTED** [review]
### Ready for Final Code Review❕
The following items need verification:
- Define the `TimeSpan.FromMilliseconds(100)` parameter as a global static property to allow advanced configuration of this constant in C#. &rarr; **Docs**: [Use time-out values](https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices-regex#use-time-out-values)
- Convert all `Regex` definitions to `GeneratedRegex`, following best practices in .NET 7+. See the `Ocelot.DependencyInjection.ConfigurationBuilderExtensions` class for an example, particularly the `SubConfigRegex()` method &rarr;
  https://github.com/ThreeMammals/Ocelot/blob/d310508aa28997a492eea4bc5cff889eb43014dd/src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs#L18-L24
  The preprocessor `if`-directive is necessary to accommodate support for `net6.0`, as well as for `net7.0` and `net8.0`
- Ensure that all `Regex` expressions are converted throughout the entire solution.

**Comment by raman-m** [comment]
@EngRajabi Hello, Mohsen! Are you energized and ready to deliver this PR? 😄

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs (hunk 12 above)
Comment (raman-m) [reply src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1348/commits/3c7781e7ceccf0a54e1feae91a07c78e12bcfa3d ✔️

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs (hunk 11 above)
Comment (raman-m) [reply src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1348/commits/3c7781e7ceccf0a54e1feae91a07c78e12bcfa3d ✔️ 
The version from the develop branch was utilized.

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs (hunk 10 above)
Comment (raman-m) [reply src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/commit/3c7781e7ceccf0a54e1feae91a07c78e12bcfa3d ✔️

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs (hunk 8 above)
Comment (raman-m) [reply src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/commit/3c7781e7ceccf0a54e1feae91a07c78e12bcfa3d ✔️

File: src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs (hunk 5 above)
Comment (raman-m) [reply src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs]: `AddOcelotBaseUrl` is `Obsolete`, so this is wrong location to handle the work.
Should be removed!

Fixed in https://github.com/ThreeMammals/Ocelot/pull/1348/commits/364d1b63ee39370f53c5e96bce2395d92b068be5 ✔️

File: src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs (hunk 6 above)
Comment (raman-m) [reply src/Ocelot/DependencyInjection/ConfigurationBuilderExtensions.cs]: `100` hardcoded value should be removed in favor of a static property!

Fixed in https://github.com/ThreeMammals/Ocelot/pull/1348/commits/364d1b63ee39370f53c5e96bce2395d92b068be5 ✔️

File: src/Ocelot/Configuration/Validator/RouteFluentValidator.cs (hunk 3 above)
Comment (raman-m) [reply src/Ocelot/Configuration/Validator/RouteFluentValidator.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1348/commits/68ef3a51989074f8ca6da43269d6ec948ceb6544 ✔️

**Review by raman-m — APPROVED** [review]
### Ready for Delivery ✅

**Comment by ggnaegi** [comment]
Sorry, but I would have first moved to .NET 7 - 9 before... @RaynaldM your Thoughts?

**Comment by ggnaegi** [comment]
@raman @RaynaldM Again, I would have prefered moving to .NET 7-9 before to reduce the complexity and ensure a better maintainability.

**Comment by raman-m** [comment]
> @ggnaegi commented [Nov 16, 2024](https://github.com/ThreeMammals/Ocelot/pull/1348#issuecomment-2480759693)

Please update to the desired .NET version in your forked repository.
To minimize complexity and ensure better maintainability, we adhere to the standard Microsoft .NET support lifecycle, specifically the [.NET and .NET Core Support Policy](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core) &rarr;  **.NET 7** &rarr; **End of support** on May 14, 2024❗ 
Our official end of .NET 6, 7 support is merged PRs #2220 #2221 &rarr; version [23.4.2](https://github.com/ThreeMammals/Ocelot/releases/tag/23.4.2)

**P.S.** Ocelot is an open-source project, just like .NET framework. You would be correct if Ocelot were a commercial project, but it is not.

**Pull request merged**